resources/migrations/000_migrations.yaml

```diff
@@ -8464,6 +8464,20 @@ databaseChangeLog:
                 UPDATE setting SET value='simple' WHERE key='humanization-strategy'
                 AND value='advanced'
 
+  - changeSet:
+      id: 312
+      author: noahmoss
+      comment: Added 0.41.0 Backfill collection_id for dashboard subscriptions
+      changes:
+        - sql:
+            sql: >-
+              UPDATE pulse p
+              SET collection_id = (
+              SELECT d.collection_id
+                FROM report_dashboard d
+                WHERE d.id = p.dashboard_id
+              );
```

Comment:
I looks like this migration is going to update every Pulse regardless of whether they have a Dashboard ID... we could make it more efficient by only updating ones with a `dashboard_id`. I think it's also going to do a separate `SELECT` to determine the appropriate `collection_id` for every single Pulse, which could get kinda slow if someone has hundreds or thousands of Pulses. Probably unlikely, but you could write a faster statement doing `UPDATE ... FROM ... WHERE` which will just join the Dashboard table once , e.g.

```sql
EXPLAIN UPDATE pulse p 
SET collection_id = d.collection_id 
FROM report_dashboard d 
WHERE p.dashboard_id = d.id 
  AND p.dashboard_id IS NOT NULL;
```

`EXPLAIN` confirms it:

```
metabase=# EXPLAIN UPDATE pulse p 
SET collection_id = (SELECT d.collection_id FROM report_dashboard d WHERE d.id = p.dashboard_id);

                                   QUERY PLAN
--------------------------------------------------------------------------------
 Update on pulse p  (cost=0.00..383.70 rows=370 width=201)
   ->  Seq Scan on pulse p  (cost=0.00..383.70 rows=370 width=201)
         SubPlan 1
           ->  Seq Scan on report_dashboard d  (cost=0.00..1.00 rows=1 width=4)
                 Filter: (id = p.dashboard_id)
(5 rows)

metabase=# EXPLAIN UPDATE pulse p 
SET collection_id = d.collection_id 
FROM report_dashboard d 
WHERE p.dashboard_id = d.id 
  AND p.dashboard_id IS NOT NULL;

                                  QUERY PLAN
-------------------------------------------------------------------------------
 Update on pulse p  (cost=0.00..14.71 rows=1 width=207)
   ->  Nested Loop  (cost=0.00..14.71 rows=1 width=207)
         Join Filter: (p.dashboard_id = d.id)
         ->  Seq Scan on pulse p  (cost=0.00..13.70 rows=1 width=193)
               Filter: (dashboard_id IS NOT NULL)
         ->  Seq Scan on report_dashboard d  (cost=0.00..1.00 rows=1 width=14)
(6 rows)
```

Reply:
Ah yeah, good catch, thanks